### ISSUE TITLE
render does not reject when rendering fails

### ISSUE THREAD

**Issue by ef4** [emberjs/ember-test-helpers#310]
I would have expected this test to pass, but it fails with "expected not to get here" in addition to an out-of-band qunit failure for the actual `Compiler Error: not-a-component is not a helper`.

```js
test('render can throw', async function(assert) {
  try {
    await render(hbs`{{not-a-component x="y"}}`);
  } catch (err) {
    assert.ok(/not-a-component is not a helper/.test(err.message), "expected to see failure");
  }
  throw new Error("expected not to get here");
});

```

**Comment by rwjblue** [comment]
This is ultimately due to how rendering is deferred in the render queue. 

There are work arounds / solutions, I’ll try to list them so we can discuss the best path forward.

**Comment by ef4** [comment]
Yeah, I know the scheduling isn't synchronous, I was just a little surprised because we're already waiting here for everything to finish, so most of the plumbing for getting the state threaded back to us must already be present.

**Comment by cibernox** [comment]
I'm also trying to find a reliable way of testing errors/assertions during the render. What are the possible workarounds for this? (Or reliable fix for that matter)

**Comment by kellyselden** [comment]
This appears to be a bit stale, but I would love some workarounds as well.

**Comment by ef4** [comment]
I still think this is a bug. `render` should throw.

**Comment by ef4** [comment]
But thanks for sharing the workaround!

**Comment by rwjblue** [comment]
> I still think this is a bug. render should throw.

Agreed. Though the actual details of how to implement this properly are non-trivial. I will try to gather better info and report here...

**Comment by Frozenfire92** [comment]
:+1: Chiming in as someone who is being affected by this

Ember 3.3

Noting neither [this workaround](https://discuss.emberjs.com/t/how-to-catch-errors-in-component-rendering-test/14854/2) or the one @kellyselden mentioned above worked for me.

Noting I'm coming to this not when rendering fails, but when a click throws an error, but assuming its the same issue with the render queue

Test
```js
test('Handles not having action passed in', async function(assert) {
    await render(hbs`{{async-button}}`);

    assert.throws(
      async () => {
        await click(find('.async-button'));
      },
      'Throws exception when action not passed in'
    );
  });
```

Relevant piece of component (within click handler)
```js
else {
  throw new Error('missing passed in action to async-button');
}
```

How it looks in qunit browser window
![screenshot from 2018-08-02 15-49-51](https://user-images.githubusercontent.com/5893112/43605347-67f5d9b6-966e-11e8-82d5-173634c45e8d.png)

**Comment by nlfurniss** [comment]
This [twiddle reproduces](https://ember-twiddle.com/de9115316563715af87fec8ddec9f796?openFiles=tests.integration.components.my-c-test.js%2Ctemplates.components.my-c.hbs) the issue.

### TLDR
There is a component that will throw an error if two attrs arent passed in (`x` and `y`) and the component is clicked on.

There are three tests:
1. Ensure no errors if both attrs passed in and component is clicked ✅
2. Ensure an error is throw is only one attr is passed in and component is clicked (**uses native JS and tests fails**) ❌
3. Ensure an error is throw is only one attr is passed in and component is clicked (uses jquery and test passes) ✅

**Comment by Frozenfire92** [comment]
@nlfurniss switching to jquery worked for me, not ideal but a temp fix until this issue is resolved, thanks!

**Comment by barryofguilder** [comment]
Are there any updates to this? I was going through a bunch of my tests where they referenced checking up on this issue. Just wanted to make sure there are still no other options than what's mentioned here.

**Comment by MrChocolatine** [comment]
Any news about this? After more than 2 years?

**Comment by buschtoens** [comment]
I've started a PR to finally fix this: [ember-test-helpers#1194](https://github.com/emberjs/ember-test-helpers/pull/1194). I would appreciate your thoughts!
